<h3 style ="text-decoration: underline"><center>Brief project: Sondage Thanksgiving</h3>

### Partie 1 : Filtrer et explorer les tendances

### 1.1. Introduction au dataset

In [9]:
# Utilisation de la fonction .read_csv() pour lire le fichier thanksgiving.csv
import pandas as pd 
data = pd.read_csv("thanksgiving.csv", encoding='latin-1')

# On affiche les 5 premières lignes du dataframe grâce à la méthode .head(5)
data.head(5)

RespondentID Do you celebrate Thanksgiving?  \
0    4337954960                            Yes   
1    4337951949                            Yes   
2    4337935621                            Yes   
3    4337933040                            Yes   
4    4337931983                            Yes   

  What is typically the main dish at your Thanksgiving dinner?  \
0                                             Turkey             
1                                             Turkey             
2                                             Turkey             
3                                             Turkey             
4                                           Tofurkey             

  What is typically the main dish at your Thanksgiving dinner? - Other (please specify)  \
0                                                NaN                                      
1                                                NaN                                      
2                                                NaN                                      
3                                                NaN                                      
4                                                NaN                                      

  How is the main dish typically cooked?  \
0                                  Baked   
1                                  Baked   
2                                Roasted   
3                                  Baked   
4                                  Baked   

  How is the main dish typically cooked? - Other (please specify)  \
0                                                NaN                
1                                                NaN                
2                                                NaN                
3                                                NaN                
4                                                NaN                

  What kind of stuffing/dressing do you typically have?  \
0                                        Bread-based      
1                                        Bread-based      
2                                         Rice-based      
3                                        Bread-based      
4                                        Bread-based      

  What kind of stuffing/dressing do you typically have? - Other (please specify)  \
0                                                NaN                               
1                                                NaN                               
2                                                NaN                               
3                                                NaN                               
4                                                NaN                               

  What type of cranberry saucedo you typically have?  \
0                                               None   
1                             Other (please specify)   
2                                           Homemade   
3                                           Homemade   
4                                             Canned   

  What type of cranberry saucedo you typically have? - Other (please specify)  \
0                                                NaN                            
1                    Homemade cranberry gelatin ring                            
2                                                NaN                            
3                                                NaN                            
4                                                NaN                            

   ...  \
0  ...   
1  ...   
2  ...   
3  ...   
4  ...   

  Have you ever tried to meet up with hometown friends on Thanksgiving night?  \
0                                                Yes                            
1                                                 No                            
2                                                Yes                            
3        

In [10]:
# Utilisation de l'attribut columns pour afficher le nom des colonnes
data.columns

Index(['RespondentID', 'Do you celebrate Thanksgiving?',
       'What is typically the main dish at your Thanksgiving dinner?',
       'What is typically the main dish at your Thanksgiving dinner? - Other (please specify)',
       'How is the main dish typically cooked?',
       'How is the main dish typically cooked? - Other (please specify)',
       'What kind of stuffing/dressing do you typically have?',
       'What kind of stuffing/dressing do you typically have? - Other (please specify)',
       'What type of cranberry saucedo you typically have?',
       'What type of cranberry saucedo you typically have? - Other (please specify)',
       'Do you typically have gravy?',
       'Which of these side dishes aretypically served at your Thanksgiving dinner? Please select all that apply. - Brussel sprouts',
       'Which of these side dishes aretypically served at your Thanksgiving dinner? Please select all that apply. - Carrots',
       'Which of these side dishes aretypically served

### 1.2. Filtrer des données

In [11]:
# On utilise la méthode value_counts() qui nous informe que 980 personnes sur les 1058
# fêtent effectivement Thanksgiving
data['Do you celebrate Thanksgiving?'].value_counts()

Yes    980
No      78
Name: Do you celebrate Thanksgiving?, dtype: int64

In [12]:
# Création d'un masque qui sélectionnera uniquement les personnes célébrant 
# effectivement Thanksgiving : On passe de 1058 entrées à 980 entrées (lignes).
masque_celeb = data['Do you celebrate Thanksgiving?']=='Yes'
data = data[masque_celeb]
data

RespondentID Do you celebrate Thanksgiving?  \
0       4337954960                            Yes   
1       4337951949                            Yes   
2       4337935621                            Yes   
3       4337933040                            Yes   
4       4337931983                            Yes   
...            ...                            ...   
1053    4335944082                            Yes   
1054    4335943173                            Yes   
1055    4335943060                            Yes   
1056    4335934708                            Yes   
1057    4335894916                            Yes   

     What is typically the main dish at your Thanksgiving dinner?  \
0                                                Turkey             
1                                                Turkey             
2                                                Turkey             
3                                                Turkey             
4                                              Tofurkey             
...                                                 ...             
1053                                             Turkey             
1054                                             Turkey             
1055                             Other (please specify)             
1056                                             Turkey             
1057                                             Turkey             

     What is typically the main dish at your Thanksgiving dinner? - Other (please specify)  \
0                                                   NaN                                      
1                                                   NaN                                      
2                                                   NaN                                      
3                                                   NaN                                      
4                                                   NaN                                      
...                                                 ...                                      
1053                                                NaN                                      
1054                                                NaN                                      
1055                                               Duck                                      
1056                                                NaN                                      
1057                                                NaN                                      

     How is the main dish typically cooked?  \
0                                     Baked   
1                                     Baked   
2                                   Roasted   
3                                     Baked   
4                                     Baked   
...                                     ...   
1053                                Roasted   
1054                                  Baked   
1055                                  Baked   
1056                                  Baked   
1057                                  Baked   

     How is the main dish typically cooked? - Other (please specify)  \
0                                                   NaN                
1                                                   NaN                
2                                                   NaN                
3                                                   NaN                
4                                                   NaN                
...                                                 ...                
1053                                                NaN                
1054                                                NaN                
1055                                                NaN                
1056                                                NaN                
1057                                                Na

### 1.3. Exploration des repas de Thanksgiving

In [13]:
# On utilise la méthode value_counts() pour connaitre les différents plats possible à Thanksgiving
data["What is typically the main dish at your Thanksgiving dinner?"].value_counts()

Turkey                    859
Other (please specify)     35
Ham/Pork                   29
Tofurkey                   20
Chicken                    12
Roast beef                 11
I don't know                5
Turducken                   3
Name: What is typically the main dish at your Thanksgiving dinner?, dtype: int64

Conclusion : La majorité des personnes(859 personnes) mangent de la dinde à Thanksgiving mais des plats plus originaux sont aussi dégustés telle que la dinde au tofu. 

In [14]:
# On créer un masque pour sélectionner uniquement les personnes mangeant du Tofurkey 
masque_tofurkey = data["What is typically the main dish at your Thanksgiving dinner?"] == 'Tofurkey'

# On applique le masque sur l'ensemble du dataset 
tofurkey = data[masque_tofurkey]

# On affiche la réponse des personnes consommant du Tofurkey concernant leur consommation de sauce
tofurkey['Do you typically have gravy?']

4      Yes
33     Yes
69      No
72      No
77     Yes
145    Yes
175    Yes
218     No
243    Yes
275     No
393    Yes
399    Yes
571    Yes
594    Yes
628     No
774     No
820     No
837    Yes
860     No
953    Yes
Name: Do you typically have gravy?, dtype: object

Conclusion: sur 20 personnes mangeant du Tofurkey, 12 en mangent avec de la sauce.

### 1.4. Tendance des desserts pour Thanksgiving

In [15]:
# Pour chacun des ingrédients: pomme, citrouille et pécan
# On utilise la fonction isna() pour voir les valeurs qui sont nulles
apple_isnull = data["Which type of pie is typically served at your Thanksgiving dinner? Please select all that apply. - Apple"].isna()
pumpkin_isnull = data["Which type of pie is typically served at your Thanksgiving dinner? Please select all that apply. - Pumpkin"].isna()
pecan_isnull = data["Which type of pie is typically served at your Thanksgiving dinner? Please select all that apply. - Pecan"].isna()

In [16]:
# On utilise l'opérateur logique & qui permet d'applique le AND sur chacun des éléments
# Cela signifie que 104 personnes n'ont mangé aucune des 3 tartes.
pies = apple_isnull & pumpkin_isnull & pecan_isnull
pies.value_counts()

False    876
True     104
dtype: int64

Conclusion: Sur les 980 personnes qui ont fêté Thanksgiving, 876 ont mangé au moins une tarte et 104 n'en ont mangé aucune. 

### Partie 2 : Liens et corrélations entre variables  

### 2.1. Convertir l'âge en une valeur numérique

In [17]:
# On crée une copy de data afin de ne pas altérer le fichier original
data2 = data.copy()

In [18]:
# On crée une fonction qui prend en paramètre d'entrée une chaine de caractères
def conversion(agestr): 
    
    # Si la chaine de caractère est nulle alors on remplace la valeur de la ligne par None
    if pd.isnull(agestr): 
        return None
    
    # On crée une liste qui contiendra les différents éléments du split de la colonne "Age"
    liste = agestr.split()
    
    # On récupère le premier élément du split
    agestr = liste[0]
    
    # Si ce premier élément contient un + comme 60+, celui-ci est remplacé par une chaine de caractère vide
    agestr = agestr.replace("+", "")
    
    # On convertit la chaine de caractères en integer afin de pouvoir faire des statistiques plus tard
    agestr = int(agestr)
    
    return agestr

In [19]:
# On utilise la méthode .apply() afin d'appliquer la fonction conversion sur l'ensemble des lignes de la colonne "Age"
# On assigne le résultat à une nouvelle colonne int_age
data2["int_age"] = data2["Age"].apply(conversion)

In [20]:
# On applique la méthode .describe() afin d'obtenir des statistiques descriptives sur la colonne "int_age"
data2["int_age"].describe()

count    947.000000
mean      40.089757
std       15.352014
min       18.000000
25%       30.000000
50%       45.000000
75%       60.000000
max       60.000000
Name: int_age, dtype: float64

Conclusion:
- On fait un décompte de l'âge sur 947 personnes dont la moyenne d'âge est de 40,1 ans avec un écart-type de 15,4ans. 
- Le plus jeune participant à 18ans et le plus vieux 60ans. 
- 50% des participants ont 45ans et moins. 

### 2.2 Convertir les revenus en valeurs numériques

In [21]:
# On crée une fonction qui prend en paramètre d'entrée une chaine de caractères
def money_convert(moneystr):
    
    # Si la chaine de caractère est nulle alors on remplace la valeur de la ligne par None
    if pd.isnull(moneystr): 
        return None
    
    # On crée une liste qui contiendra les différents éléments du split de la colonne répertoriant les revenus annuels communs
    liste = moneystr.split()
    
    # On récupère le premier élément du split 
    moneystr = liste[0]
    
    # Si ce premier élément est "Prefer" on retourne None
    if moneystr == "Prefer": 
        return None
    
    # On remplace les signes $ ou , par une chaine de caractères vide
    # On utilise les expressions régulières 
    moneystr = moneystr.replace('$',"").replace(',',"")
   
    # On convertit la chaine de caractères en integer afin de pouvoir faire des statistiques plus tard
    moneystr = int(moneystr)
    
    return moneystr

In [22]:
# On utilise la méthode .apply() afin d'appliquer la fonction money_convert sur l'ensemble des lignes de la colonne d'intérêt
# On assigne le résultat à une nouvelle colonne int_income
# Par défaut la fonction apply s'applique sur la colonne
data2["int_income"] = data2["How much total combined money did all members of your HOUSEHOLD earn last year?"].apply(money_convert)

In [23]:
# On applique la méthode .describe() afin d'obtenir des statistiques descriptives sur la colonne "int_income"
data2["int_income"].describe()

count       829.000000
mean      75965.018094
std       59068.636748
min           0.000000
25%       25000.000000
50%       75000.000000
75%      100000.000000
max      200000.000000
Name: int_income, dtype: float64

Conclusion:
- On fait un décompte du revenu sur 829 personnes dont le revenu moyen est de 75 965 dollars avec un écart-type de 59 069 dollars. 
- Le revenu le plus faible est de 0 dollars et le plus élevé est de 200 000 dollars. 
- 50% des participants ont des revenus de 75 000 dollars et moins. 

### 2.3 Corrélation entre distance et revenus

In [24]:
# On crée un masque qui retournera un booléen True si les valeurs dans la colonne int_income sont inférieures à 150000
masque_inf_150K = data2["int_income"] < 150000

# On applique le masque à l'ensemble du jeu de données pour récupérer les valeurs adéquates
salaire_bas = data2[masque_inf_150K]

In [25]:
# On utilise la méthode .value_counts() pour connaitre les habitudes de voyages pour Thanksgiving
# Ainsi, on sait que parmi les familles gagnant moins de 150K, 281 ne voyagent pas du tout à Thanksgiving.
salaire_bas["How far will you travel for Thanksgiving?"].value_counts()

Thanksgiving is happening at my home--I won't travel at all                         281
Thanksgiving is local--it will take place in the town I live in                     203
Thanksgiving is out of town but not too far--it's a drive of a few hours or less    150
Thanksgiving is out of town and far away--I have to drive several hours or fly       55
Name: How far will you travel for Thanksgiving?, dtype: int64

Conclusion : Parmi les personnes avec moins de 150 000 dollars de revenus, 281 ne voyagent pas du tout à Thanksgiving tandis que 55 voyagent en dehors de la ville.

In [26]:
# On applique la même méthode que précédemment pour les familles gagnant plus de 150K
masque_sup_150K = data2["int_income"] >= 150000
salaire_haut = data2[masque_sup_150K]

In [27]:
# Ainsi, on sait que parmi les familles gagnant plus de 150K, 66 ne voyagent pas du tout à Thanksgiving.
salaire_haut["How far will you travel for Thanksgiving?"].value_counts()

Thanksgiving is happening at my home--I won't travel at all                         66
Thanksgiving is local--it will take place in the town I live in                     34
Thanksgiving is out of town but not too far--it's a drive of a few hours or less    25
Thanksgiving is out of town and far away--I have to drive several hours or fly      15
Name: How far will you travel for Thanksgiving?, dtype: int64

Conclusion : Parmi les personnes avec plus de 150 000 dollars de revenus, 66 ne voyagent pas du tout à Thanksgiving tandis que 15 voyagent en dehors de la ville.

### 2.4 Lien entre passer Thanksgiving entre amis avec l'âge et le revenu

In [28]:
# On utilise la méthode .pivot_table() qui permet de passer en index les valeurs uniques des 2 questions
# En colonnes on définit les valeurs comme étant celle de la colonne "int_age" du dataset

# On sait donc que la moyenne d'âge des personnes ayant célébré un Friendsgiving est de 37ans 
# On sait que la moyenne d'âge des personnes voulant voir leur amis d'enfance la nuit de Thanksgiving est de 34ans
pivot_table1 = data2.pivot_table(index = "Have you ever tried to meet up with hometown friends on Thanksgiving night?", values ="int_age", columns = 'Have you ever attended a "Friendsgiving?"')
print(pivot_table1)

Have you ever attended a "Friendsgiving?"                  No        Yes
Have you ever tried to meet up with hometown fr...                      
No                                                  42.283702  37.010526
Yes                                                 41.475410  33.976744


In [134]:
# On utilise la méthode .pivot_table() qui permet de passer en index les valeurs uniques des 2 questions
# En colonnes on définit les valeurs comme étant celle de la colonne "int_income" du dataset

# On sait donc que le revenu moyen des personnes ayant célébré un Friendsgiving est de $72,895 
# On sait que le revenu moyen des personnes voulant voir leur amis d'enfance la nuit de Thanksgiving est de $66,020
pivot_income = data2.pivot_table(index = "Have you ever tried to meet up with hometown friends on Thanksgiving night?", values ="int_income", columns = 'Have you ever attended a "Friendsgiving?"')
print(pivot_income)

Have you ever attended a "Friendsgiving?"                     No           Yes
Have you ever tried to meet up with hometown fr...                            
No                                                  78914.549654  72894.736842
Yes                                                 78750.000000  66019.736842
